akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/BySliceSourceProviderAdapter.scala

```diff
@@ -6,42 +6,77 @@ package akka.projection.r2dbc.internal
 
 import java.time.Instant
 import java.util.Optional
+import java.util.concurrent.CompletableFuture
 import java.util.concurrent.CompletionStage
 import java.util.function.Supplier
+import java.util.function.{ Function => JFunction }
+
+import scala.compat.java8.FutureConverters._
+import scala.compat.java8.OptionConverters._
 import scala.concurrent.Future
+
 import akka.NotUsed
 import akka.annotation.InternalApi
 import akka.dispatch.ExecutionContexts
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.query.typed.scaladsl.EventTimestampQuery
+import akka.persistence.query.typed.scaladsl.LoadEventQuery
+import akka.projection.BySlicesSourceProvider
+import akka.projection.internal.SourceProviderAdapter
 import akka.projection.javadsl
+import akka.projection.javadsl.CustomStartOffsetSourceProvider
 import akka.projection.scaladsl
 import akka.stream.scaladsl.Source
 
-import scala.compat.java8.FutureConverters._
+@InternalApi private[projection] object BySliceSourceProviderAdapter {
+  def apply[Offset, Envelope](
+      sourceProvider: javadsl.SourceProvider[Offset, Envelope]): scaladsl.SourceProvider[Offset, Envelope] =
+    sourceProvider match {
```

Comment:
this javadsl -> scaladsl adapter is still rather messy, but need to unwrap the CustomStartOffsetSourceProvider decoration and implement it here instead to keep the BySlicesSourceProvider type

Reply:
Actually, it will be a problem also for the normal scaladsl case. I'll have to take another look at that.